waiter/src/waiter/auth/composite.clj

```diff
@@ -0,0 +1,51 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.auth.composite
+  (:require [plumbing.core :as pc]
+            [waiter.auth.authentication :as auth]
+            [waiter.auth.saml :as saml]
+            [waiter.util.utils :as utils]))
+
+(defrecord CompositeAuthenticator [authenticators saml-acs-handler-fn saml-auth-redirect-handler-fn]
+  auth/Authenticator
+  (wrap-auth-handler [_ request-handler]
+    (let [handlers (pc/map-vals #(auth/wrap-auth-handler % request-handler) authenticators)]
+      (fn composite-authenticator-handler [request]
+        (let [authentication (get-in request [:waiter-discovery :service-parameter-template "authentication"])]
+          (if authentication
+            (if-let [handler ((keyword authentication) handlers)]
```

Comment:
Nitpick: I find `(get handlers (keyword authentication))` easier to read

Reply:
done